### ISSUE TITLE
ASN1_digest vs2015 cast error

### ISSUE THREAD

**Issue by grunk** [civetweb/civetweb#425]
Hello,

i just updated from civetweb 1.8 to 1.9.1 in my project and i'm getting a cast error on : 
`ASN1_digest((int (*)())i2d_X509, digest, (char *)cert, buf, &ulen);` 
https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L12255

> Error	C2664	'int ASN1_digest(i2d_of_void (__cdecl *),const EVP_MD *,char *,unsigned char *,unsigned int *)': cannot convert argument 1 from 'int (__cdecl *)(void)' to 'i2d_of_void (__cdecl *)'	vendor	D:\C++\Projets\v8\src\cpp\vendor\civet\civetweb.cpp	11859

I managed to compile my project and running civetweb by replacing the line by : 
`ASN1_digest((i2d_of_void *)i2d_X509, digest, (char *)cert, buf, &ulen);`

My handlers never seem to use this part of code , so i don't really know if this fix is correct or not.

I'm compiling with NO_SSL_DL , so maybe this is part of the problem.

**Comment by bel2125** [comment]
I think you get this error, because you renamed civetweb.c to civetweb.cpp, forcing to compile as C++.
In C, a function `int (*)()` may have as many arguments as you like, in C++ `int (*)()` has no arguments (corresponding to `int (*)(void)` in C). Since you compile with C++, you get a compile error, and I don't get it by compiling with C.

If this is the case, the solution should be to replace
```C
#define ASN1_digest                                                            \
	(*(int (*)(int (*)(),                                                      \
	           const EVP_MD *,                                                 \
	           char *,                                                         \
	           unsigned char *,                                                \
	           unsigned int *))crypto_sw[10].ptr)
 ```
by
```C
#define ASN1_digest                                                            \
	(*(int (*)(int (*)(void*, unsigned char **),                                                      \
	           const EVP_MD *,                                                 \
	           char *,                                                         \
	           unsigned char *,                                                \
	           unsigned int *))crypto_sw[10].ptr)
 ```
and
```C
ASN1_digest((int (*)())i2d_X509, digest, (char *)cert, buf, &ulen);
```
by
```C
ASN1_digest((int (*)(void*, unsigned char **))i2d_X509, digest, (char *)cert, buf, &ulen);
```

Although it is not required in C (only in C++), it does not hurt in C either. So I would just change this in civetweb.c and run all the automatic builds again. Would you test it in your C++ case as well?

**Comment by grunk** [comment]
Yes indeed , I renamed civetweb.c to cpp . I'll try your fix tomorrow and let you know.

**Comment by bel2125** [comment]
While in 1.9.1, compiling with a C++ compiler was not an issue, recently some changes were added to allow compiling as C++ (https://github.com/civetweb/civetweb/pull/418). Yes, more typecasts would be worth a pull request, but please make sure you use the current head version as a base, not the 1.9.1 release version.

**Comment by grunk** [comment]
Using the Head of master fix all the cast error i had. Your fix in 8296b4298ea3aef07c709ea700527bfa7f66bba4 seems to works well too
